### ISSUE TITLE
What causes "Connection refused: no further information"?

### ISSUE THREAD

**Issue by cowwoc** [eclipse/jetty.project#8720]
**Jetty version**
11.0.12

**Java version**
JDK 19.0

**Environment**
Microsoft Windows [Version 10.0.19044.2006]

**Question**
Following up on https://github.com/eclipse/jetty.project/issues/8681, could `Connection refused: no further information` be triggered by ephemeral port exhaustion? Are there other situations that would trigger it?

**Scenario**

I am sending thousands of HTTP/1 requests  over a SOCKS proxy that runs over an SSH tunnel.

Eventually, one of the connections goes into a bad state where all HTTP requests fail with `Connection refused: no further information`. I tried hitting multiple proxies when this condition occurs, but they all return the same error message.

Strangely, if I SSH into the proxy machine myself and invoke `curl <destination>` it works just fine.

This leads me to conclude that something is wrong with:

1. The SSH connection or SOCKS proxy,
2. Jetty, or
3. The operating system

The destination server can't be at fault because `curl` works.

I am running into the same behavior when running on AWS Linux 2 so the problem is not Windows-specific.

Do you have any suggestions on how to debug this further?

**Comment by cowwoc** [comment]
Looking at some debug logs, I'm pretty sure it's an SSH problem...

There is a period where I don't use the proxies for multiple minutes. An idle timeout is triggered and the SSH tunnel closes. A few minutes later, I ask Jetty to make an HTTP request through the SOCKS proxy that ran over the closed SSH connection and I get the aforementioned error.

I'm going to go ahead and close this issue for now.

**Comment by cowwoc** [comment]
I've got a follow-up question: Is there a programmatic way to detect that the proxy that is associated with `HttpClient` is down?

**Comment by sbordet** [comment]
The proxy is a different host machine. Open a socket to it and see if you can.

**Comment by cowwoc** [comment]
@sbordet True.

But out of curiosity, does `HttpClient` know if the exception it is thrown is caused by the proxy or destination? If so, could it provide some sort of indication which case occurred (perhaps by wrapping proxy-thrown exceptions with another exception)?

**Comment by sbordet** [comment]
@cowwoc not sure I understand?

**Comment by joakime** [comment]
@cowwoc not sure that's possible, the client only has a socket connection, if a problem occurs, it's on that first socket from the point of view of the client.  

Eg: your actual path to the resource could be ...  

1. http client
2. local corporate proxy
3. destination http proxy / load balancer
4. destination http server
5. backend content system

The only knowledge / awareness the client has at the connection level is between 1 and 2. (everything after 2 doesn't look like a different connection to the client)

**Comment by cowwoc** [comment]
@joakime I can see this for data being read/written, but can a client differentiate between connections closed by the destination and proxy?

When a destination closes a connection, the proxy *also* closes the connection? Or are the two events independent?

**Comment by joakime** [comment]
If the destination closes the connection, the proxy can close it too, but interestingly it's not really mandated by the specs that the proxy closes the client side connection too.

For example, in [RFC 7230: HTTP/1.1 spec - Section 6.1 - Connection](https://www.rfc-editor.org/rfc/rfc7230#section-6.1), if the client sends `Connection: close` (or any `Connection` header values) the proxy MUST remove them before forwarding to the destination.

Basically, the two sides (client -> proxy, and proxy -> server) have different connection management rules.
But the client is rendered completely oblivious to this fact, it only knows about the client -> proxy side of the connection management.

**Comment by cowwoc** [comment]
Got it. Thank you for the detailed explanation. I will close this issue.
